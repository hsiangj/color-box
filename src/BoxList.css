.BoxList-boxes {
  justify-content: space-evenly;
}

