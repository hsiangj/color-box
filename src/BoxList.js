import React, {useState} from 'react';
import NewBoxForm from './NewBoxForm';
import Box from './Box';
import './BoxList.css';

const BoxList = () => {
  const [boxes, setBoxes] = useState([]);
  const addBox = (newBox) => {
    setBoxes(boxes => [...boxes, newBox])
  }

const remove = (id) => {
  setBoxes(boxes => boxes.filter(box => box.id !== id));
}

  return (
    <div className='BoxList'>
      <h2>Color Box Maker</h2>
      <NewBoxForm addBox={addBox} />
      <div className='BoxList-boxes'>
        {boxes.map(({id, width, height, color}) => <Box id={id} width={width} height={height} color={color} key={id} remove={remove}/>)}
      </div>
    </div>
  )
}


export default BoxList;