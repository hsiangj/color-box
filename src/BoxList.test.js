import { render, fireEvent } from '@testing-library/react';
import BoxList from './BoxList';

it("renders without crashing", function(){
  render(<BoxList />);
});

it("matches snapshot", function(){
  const {asFragment} = render(<BoxList />);
  expect(asFragment()).toMatchSnapshot();
});

function createBox(boxList, height="50", width="50", color="blue"){
  // select input fields
  const heightInput = boxList.getByLabelText('Height (10-300)');
  const widthInput = boxList.getByLabelText('Width (10-300)');
  const colorInput = boxList.getByLabelText('Color');
  // add value to input fields
  fireEvent.change(heightInput, {target: {value: height}});
  fireEvent.change(widthInput, {target: {value: width}});
  fireEvent.change(colorInput, {target: {value: color}});
  // select button
  const btn = boxList.queryByText('Create Box');
  // click button to submit input fields
  fireEvent.click(btn);
}

it("should add new box", function(){
  const boxList = render(<BoxList />);
  // no box yet therefore no remove button
  expect(boxList.queryByText('x')).not.toBeInTheDocument();
  createBox(boxList);
  const removeButton = boxList.queryByText('x');
  expect(removeButton).toBeInTheDocument();
  expect(removeButton.previousSibling).toHaveStyle("width: 50px");
  // empty form. Color picker does not show empty string
  expect(boxList.getAllByDisplayValue("")).toHaveLength(2);
  
});

it("able to remove a box", function(){
  const boxList = render(<BoxList />);
  createBox(boxList);
  const removeButton = boxList.queryByText('x');
  fireEvent.click(removeButton);
  expect(removeButton).not.toBeInTheDocument();
});