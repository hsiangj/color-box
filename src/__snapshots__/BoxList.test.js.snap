// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="BoxList"
  >
    <h2>
      Color Box Maker
    </h2>
    <form>
      <label
        for="width"
      >
        Width (10-300) 
      </label>
      <input
        id="width"
        max="300"
        min="10"
        name="width"
        type="number"
        value=""
      />
      <br />
      <label
        for="height"
      >
        Height (10-300) 
      </label>
      <input
        id="height"
        max="300"
        min="10"
        name="height"
        type="number"
        value=""
      />
      <br />
      <label
        for="color"
      >
        Color 
      </label>
      <input
        id="color"
        name="color"
        type="color"
        value=""
      />
      <br />
      <button>
        Create Box
      </button>
    </form>
    <div
      class="BoxList-boxes"
    />
  </div>
</DocumentFragment>
`;
